File: Twixodus/Pipeline/ThreadSelection.swift
// Step 4 of the pipeline: pick which threads to import for the date range.
//
// A thread whose first tweet falls inside the configured range is imported as
// usual. Threads that *started* before the range but were extended with new
// tweets inside it are a special case: Day One can't update an existing entry,
// so the whole thread has to be imported again and the older, shorter copy
// deleted by hand. This module finds those threads, and formats the reminder
// to delete the duplicates.

import Foundation

/// What a run would do with an archive, computed up front for the UI: how many
/// threads are new, how many grew since their import and need the full
/// re-import treatment, and how many are already in Day One. Mirrors exactly
/// what the engine will skip and import.
public struct ImportPreview {
    public var newThreads = 0
    public var grownThreads = 0
    public var alreadyImported = 0
    /// What the run will actually send to Day One.
    public var pending: Int { newThreads + grownThreads }
}

public enum ThreadSelection {

    /// Computes the ImportPreview for these threads against the ledger state,
    /// with the same selection rules the run itself will use.
    public static func preview(
        _ threads: [TweetThread], startDate: Date, endDate: Date,
        processedIDs: Set<String>, coveredThrough: Date?
    ) -> ImportPreview {
        let (inRange, extended) = partitionThreadsByDate(
            threads, startDate: startDate, endDate: endDate,
            processedIDs: processedIDs, coveredThrough: coveredThrough)
        let newThreads = inRange.filter { !processedIDs.contains($0[0].idStr) }.count
        let grown = extended.filter { !processedIDs.contains(ImportLedger.extensionMarker($0)) }.count
        return ImportPreview(
            newThreads: newThreads,
            grownThreads: grown,
            alreadyImported: (inRange.count - newThreads) + (extended.count - grown))
    }

    /// Splits threads into the ones to import normally and the ones to re-import.
    ///
    /// Returns (inRange, extended):
    ///   - inRange: threads whose first tweet was posted within the date range.
    ///   - extended: threads that already live in Day One as a shorter entry and
    ///     gained tweets since. Either they began before the range and grew into
    ///     it, or — when the last import's coverage point is known — they were
    ///     imported whole and their tail is newer than that point. Both kinds get
    ///     imported again in full and reported to the user.
    ///
    /// Threads that started after the range, or that ended before it, are dropped.
    public static func partitionThreadsByDate(
        _ threads: [TweetThread], startDate: Date, endDate: Date,
        processedIDs: Set<String> = [], coveredThrough: Date? = nil
    ) -> (inRange: [TweetThread], extended: [TweetThread]) {
        var inRange: [TweetThread] = []
        var extended: [TweetThread] = []

        for thread in threads {
            guard let root = thread.first else { continue }
            let rootDate = root.createdAt

            if startDate <= rootDate && rootDate <= endDate {
                // An already-imported thread whose tail is newer than what the
                // last import covered has grown since — Day One can't extend
                // the existing entry, so it needs the full re-import treatment
                // even though its root falls inside the range.
                if let covered = coveredThrough, rootDate <= covered,
                   processedIDs.contains(root.idStr),
                   thread.contains(where: { covered < $0.createdAt && $0.createdAt <= endDate }) {
                    extended.append(thread)
                } else {
                    inRange.append(thread)
                }
            } else if rootDate < startDate,
                      thread.dropFirst().contains(where: { startDate <= $0.createdAt && $0.createdAt <= endDate }) {
                extended.append(thread)
            }
        }

        return (inRange, extended)
    }

    /// Debug helper: keeps only the threads containing any of the listed
    /// tweet IDs. (The Python version matched root IDs only; matching any
    /// tweet is friendlier when the interesting tweet sits mid-thread.)
    public static func filterToDebugIDs(_ threads: [TweetThread], ids: Set<String>) -> [TweetThread] {
        threads.filter { thread in thread.contains { ids.contains($0.idStr) } }
    }

    /// Counts the tweets the already-imported copy of this thread is expected
    /// to hold: everything before the range — or, when the last import's
    /// coverage point is known and the thread existed by then, everything up
    /// to that point (a completed import always took the whole thread as the
    /// archive had it).
    public static func countTweetsBefore(
        _ thread: TweetThread, startDate: Date, coveredThrough: Date? = nil
    ) -> Int {
        let beforeRange = thread.filter { $0.createdAt < startDate }.count
        guard let covered = coveredThrough,
              let rootDate = thread.first?.createdAt, rootDate <= covered
        else { return beforeRange }
        return max(beforeRange, thread.filter { $0.createdAt <= covered }.count)
    }

    /// Builds a link to a tweet, falling back to the account-agnostic URL.
    public static func tweetURL(tweetId: String, username: String?, host: String = "twitter.com") -> String {
        if let username, !username.isEmpty {
            return "https://\(host)/\(username)/status/\(tweetId)"
        }
        return "https://\(host)/i/web/status/\(tweetId)"
    }

    private static let reportDateFormatter = PipelineDates.formatter("dd MMMM yyyy, HH:mm")

    /// Formats the list of re-imported threads into a reminder to delete the
    /// older duplicates.
    public static func formatReimportReport(
        _ reimported: [ImportedEntry], username: String?, linkHost: String = "twitter.com"
    ) -> String {
        let line = String(repeating: "=", count: 72)
        let header = """
            \(line)
            ⚠️  ACTION REQUIRED — \(reimported.count) thread(s) were re-imported in full
            \(line)
            These threads were imported once before, but you added more tweets to
            them afterwards. Day One can't extend an existing entry, so each one was
            imported again as a complete thread — which leaves an older, shorter
            copy in your journal. Delete the older copy of each:

            """

        var blocks: [String] = []
        for (i, entry) in reimported.enumerated() {
            let dateStr = reportDateFormatter.string(from: entry.date)
            let oldEntryStr: String
            if let count = entry.previousTweetCount, count != 0 {
                oldEntryStr = "≈\(count) tweets (the shorter, previously imported copy)"
            } else {
                oldEntryStr = "the shorter, previously imported copy"
            }
            blocks.append("""

                \(i + 1). \(dateStr) — “\(entry.title)”
                   Journal: \(entry.journal)
                   Old entry to delete: \(oldEntryStr)
                   New entry: \(entry.tweetCount) tweets
                   First tweet: \(tweetURL(tweetId: entry.tweetId, username: username, host: linkHost))

                """)
        }

        let footer = """

            Both copies share the same entry date, so searching that date in Day One
            will show them side by side — keep the longer one.

            """

        return header + blocks.joined() + footer
    }
}

File: thread_selection.py
"""Step 4 of the pipeline: pick which threads to import for the date range.

A thread whose first tweet falls inside the configured range is imported as
usual. Threads that *started* before the range but were extended with new
tweets inside it are a special case: Day One can't update an existing entry,
so the whole thread has to be imported again and the older, shorter copy
deleted by hand. This module finds those threads, and formats + saves the
reminder to delete the duplicates.
"""

from __future__ import annotations

from datetime import datetime
from pathlib import Path
from typing import List, Optional, Tuple

from models import ImportedEntry, Thread

REPORT_PATH = str(Path(__file__).resolve().parent / "threads_to_delete.txt")


def parse_date_range(start_date_str: str, end_date_str: str):
    """Parses the "DD Month YYYY" dates from config. Raises ValueError if malformed."""
    start_date = datetime.strptime(start_date_str, "%d %B %Y")
    end_date = datetime.strptime(end_date_str, "%d %B %Y")
    return start_date, end_date


def _thread_dates(thread: Thread) -> list:
    """Returns the naive creation dates of every tweet in a thread, in thread order."""
    dates = []
    for tweet_in_thread in thread:
        tweet_data = tweet_in_thread.get("tweet", {})
        created_at = tweet_data.get("created_at")
        if isinstance(created_at, datetime):
            dates.append(created_at.replace(tzinfo=None))
    return dates


def partition_threads_by_date(
    threads: List[Thread], start_date: datetime, end_date: datetime
) -> Tuple[List[Thread], List[Thread]]:
    """Splits threads into the ones to import normally and the ones to re-import.

    Returns a tuple of (in_range, extended):
      - in_range: threads whose first tweet was posted within the date range.
      - extended: threads that began before the range but gained at least one
        tweet inside it. Those already live in Day One as a shorter entry, so
        they get imported again in full and reported to the user.

    Threads that started after the range, or that ended before it, are dropped.
    """
    in_range = []
    extended = []

    for thread in threads:
        dates = _thread_dates(thread)
        if not dates:
            continue

        root_date = dates[0]
        if start_date <= root_date <= end_date:
            in_range.append(thread)
        elif root_date < start_date and any(
            start_date <= date <= end_date for date in dates[1:]
        ):
            extended.append(thread)

    return in_range, extended


def count_tweets_before(thread: Thread, start_date: datetime) -> int:
    """Counts the tweets of a thread posted before the range — i.e. how big the
    already-imported copy of this thread is expected to be."""
    return sum(1 for date in _thread_dates(thread) if date < start_date)


def tweet_url(tweet_id: str, username) -> str:
    """Builds a link to a tweet, falling back to the account-agnostic URL."""
    if username:
        return f"https://twitter.com/{username}/status/{tweet_id}"
    return f"https://twitter.com/i/web/status/{tweet_id}"


def format_reimport_report(
    reimported: List[ImportedEntry], start_date: datetime, username=None
) -> str:
    """Formats the list of re-imported threads into a reminder to delete the
    older duplicates."""
    line = "=" * 72
    header = (
        f"{line}\n"
        f"⚠️  ACTION REQUIRED — {len(reimported)} thread(s) were re-imported in full\n"
        f"{line}\n"
        "These threads were started before the current date range, but you added\n"
        "more tweets to them within it. Day One can't extend an existing entry, so\n"
        "each one was imported again as a complete thread — which leaves an older,\n"
        "shorter copy in your journal. Delete the older copy of each:\n"
    )

    blocks = []
    for i, entry in enumerate(reimported, start=1):
        date_str = entry.date.strftime("%d %B %Y, %H:%M")
        old_entry_str = (
            f"≈{entry.previous_tweet_count} tweets (those posted before "
            f"{start_date.strftime('%d %B %Y')})"
            if entry.previous_tweet_count
            else "the shorter, previously imported copy"
        )
        blocks.append(
            f"\n{i}. {date_str} — “{entry.title}”\n"
            f"   Journal: {entry.journal}\n"
            f"   Old entry to delete: {old_entry_str}\n"
            f"   New entry: {entry.tweet_count} tweets\n"
            f"   First tweet: {tweet_url(entry.tweet_id, username)}\n"
        )

    footer = (
        "\nBoth copies share the same entry date, so searching that date in Day One\n"
        "will show them side by side — keep the longer one.\n"
    )

    return header + "".join(blocks) + footer


def write_reimport_report(report: str):
    """Prints the delete-the-duplicates reminder and saves it for later reference."""
    print("\n" + report)
    with open(REPORT_PATH, "w") as f:
        f.write(report)
    print(f"Saved this reminder to {REPORT_PATH}")
